portset/portset.go

```diff
@@ -0,0 +1,225 @@
+
+	ok = false
+	key := userName + ":" + port
+
+	if _, err := p.userPortMap.Get(key); err == nil {
+		ok = true
+	} else {
+		p.userPortMap.Add(key, p)
+	}
+
+	// program the ipset
+	if err = p.addPortSet(userName, port); err != nil {
+		return false, err
+	}
+
+	return ok, nil
+}
+
+// AddUserPortSet : This adds/updates userPortSet cache. This cache
+// maps user to the portset associated with its PU. This is called during
+// creation of PU/on reception of application SYN-ACK packet.
+func (p *portSetInstance) AddUserPortSet(userName string, portset string) (err error) {
+
+	p.userPortSet.AddOrUpdate(userName, portset)
+	return nil
+
+}
+
+// GetUserPortSet returns the portset associated with user.
+func (p *portSetInstance) getUserPortSet(userName string) (port string, err error) {
+
+	portSetName, err := p.userPortSet.Get(userName)
+	if err != nil {
+		return "", err
+	}
+	return portSetName.(string), nil
+}
+
+// DelUserPortSet  deletes user from userPortSet cache.
+func (p *portSetInstance) DelUserPortSet(userName string) (err error) {
+
+	return p.userPortSet.Remove(userName)
+}
+
+// addPortSet  programs the ipset portset with port. The
+// portset name is derived from userPortSet cache.
+func (p *portSetInstance) addPortSet(userName string, port string) (err error) {
+
+	puPortSetName, err := p.getUserPortSet(userName)
+	if err != nil {
+		return fmt.Errorf("Unable to get portset from uid")
+	}
+
+	ips := ipset.IPSet{
+		Name: puPortSetName,
+	}
+
+	if _, err := strconv.Atoi(port); err != nil {
+		return fmt.Errorf("Not a valid Port")
+	}
+
+	if adderr := ips.Add(port, portEntryTimeout); adderr != nil {
+		return fmt.Errorf("Unable to add port to set")
+	}
+	return nil
+}
+
+// deletePortSet deletes the portset
+func (p *portSetInstance) deletePortSet(userName string, port string) (err error) {
+
+	puPortSetName, err := p.getUserPortSet(userName)
+	if err != nil {
+		return fmt.Errorf("Unable to get portset from uid")
+	}
+
+	ips := ipset.IPSet{
+		Name: puPortSetName,
+	}
+
+	if _, err := strconv.Atoi(port); err != nil {
+		return fmt.Errorf("Not a valid Port")
+	}
+
+	if delerr := ips.Del(port); delerr != nil {
+		return fmt.Errorf("Unable to delete port from portset")
+	}
+
+	return nil
+}
+
+// startPortSetTask This go routine periodically scans (1s)
+// /proc/net/tcp file for listening ports and programs
+// the portsets. This worker thread is setup during datapath
+// initilisation.
+func startPortSetTask(p *portSetInstance) {
+	t := time.NewTicker(portSetUpdateIntervalMilliseconds * time.Millisecond)
+	for range t.C {
+		// Update PortSet periodically.
+		p.updateIPPortSets()
+	}
+}
+
+func (p *portSetInstance) updateIPPortSets() {
+	buffer, err := ioutil.ReadFile(procNetTCPFile)
+	if err != nil {
+		zap.L().Warn("Failed to read /proc/net/tcp file", zap.Error(err))
+		// This is a go routine, cannot return error
+		return
+	}
+
+	s := string(buffer)
+	// Ignoring the last \n
+	s = s[:len(s)-1]
```

Comment:
not needed. check for empty line in for loop and skip. even if we need to do this use string.TrimSuffix.